src/strands/ir_types.js

```diff
@@ -0,0 +1,238 @@
+  PHI: 'phi',
+  STATEMENT: 'statement',
+};
+
+export const NodeTypeToName = Object.fromEntries(
+  Object.entries(NodeType).map(([key, val]) => [val, key])
+);
+
+export const NodeTypeRequiredFields = {
+  [NodeType.OPERATION]: ["opCode", "dependsOn", "dimension", "baseType"],
+  [NodeType.LITERAL]: ["value", "dimension", "baseType"],
+  [NodeType.VARIABLE]: ["identifier", "dimension", "baseType"],
+  [NodeType.CONSTANT]: ["value", "dimension", "baseType"],
+  [NodeType.STRUCT]: [""],
+  [NodeType.PHI]: ["dependsOn", "phiBlocks", "dimension", "baseType"],
+  [NodeType.STATEMENT]: ["opCode"]
+};
+
+export const StatementType = {
+  DISCARD: 'discard',
+};
+
+export const BaseType = {
+  FLOAT: "float",
+  INT: "int",
+  BOOL: "bool",
+  MAT: "mat",
+  DEFER: "defer",
+  SAMPLER2D: "sampler2D",
+};
+
+export const BasePriority = {
+  [BaseType.FLOAT]: 3,
+  [BaseType.INT]: 2,
+  [BaseType.BOOL]: 1,
+  [BaseType.MAT]: 0,
+  [BaseType.DEFER]: -1,
+  [BaseType.SAMPLER2D]: -10,
+};
+
+export const DataType = {
+  float1: { fnName: "float", baseType: BaseType.FLOAT, dimension:1, priority: 3,  },
+  float2: { fnName: "vec2", baseType: BaseType.FLOAT, dimension:2, priority: 3,  },
+  float3: { fnName: "vec3", baseType: BaseType.FLOAT, dimension:3, priority: 3,  },
+  float4: { fnName: "vec4", baseType: BaseType.FLOAT, dimension:4, priority: 3,  },
+  int1: { fnName: "int", baseType: BaseType.INT, dimension:1, priority: 2,  },
+  int2: { fnName: "ivec2", baseType: BaseType.INT, dimension:2, priority: 2,  },
+  int3: { fnName: "ivec3", baseType: BaseType.INT, dimension:3, priority: 2,  },
+  int4: { fnName: "ivec4", baseType: BaseType.INT, dimension:4, priority: 2,  },
+  bool1: { fnName: "bool", baseType: BaseType.BOOL, dimension:1, priority: 1,  },
+  bool2: { fnName: "bvec2", baseType: BaseType.BOOL, dimension:2, priority: 1,  },
+  bool3: { fnName: "bvec3", baseType: BaseType.BOOL, dimension:3, priority: 1,  },
+  bool4: { fnName: "bvec4", baseType: BaseType.BOOL, dimension:4, priority: 1,  },
+  mat2: { fnName: "mat2x2", baseType: BaseType.MAT, dimension:2, priority: 0,  },
+  mat3: { fnName: "mat3x3", baseType: BaseType.MAT, dimension:3, priority: 0,  },
+  mat4: { fnName: "mat4x4", baseType: BaseType.MAT, dimension:4, priority: 0,  },
+  defer: { fnName:  null, baseType: BaseType.DEFER, dimension: null, priority: -1 },
+  sampler2D: { fnName: "texture", baseType: BaseType.SAMPLER2D, dimension: 1, priority: -10 },
+}
+export const structType = function (hookType) {
+  // const hookType = hookType?.name ? hookType
+  let T = hookType.type === undefined ? hookType : hookType.type; 
+  const structType = { 
+    name: hookType.name,
+    properties: [],
+    typeName: T.typeName,
+  };
+  for (const prop of T.properties) {
+    const propType = TypeInfoFromGLSLName[prop.type.typeName];
+    structType.properties.push(
+      {name: prop.name, dataType: propType }
+    );
+  }
+  return structType;
+};
+
+export const StructType = {
+  Vertex: {
```

Comment:
Ok I think this is actually the last code thing we need: rather than hardcoding these, we need to map from the output of the hook info function to something in this format. There's one failing test and it's due to this